compiler/rustc_const_eval/src/interpret/terminator.rs

```diff
@@ -174,68 +173,120 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
     }
 
     fn check_argument_compat(
-        rust_abi: bool,
-        caller: TyAndLayout<'tcx>,
-        callee: TyAndLayout<'tcx>,
+        caller_abi: &ArgAbi<'tcx, Ty<'tcx>>,
+        callee_abi: &ArgAbi<'tcx, Ty<'tcx>>,
     ) -> bool {
-        if caller.ty == callee.ty {
-            // No question
+        // Heuristic for type comparison.
+        let layout_compat = || {
+            if caller_abi.layout.ty == callee_abi.layout.ty {
+                // No question
+                return true;
+            }
+            // Compare layout
+            match (caller_abi.layout.abi, callee_abi.layout.abi) {
+                // Different valid ranges are okay (once we enforce validity,
+                // that will take care to make it UB to leave the range, just
+                // like for transmute).
+                (abi::Abi::Scalar(caller), abi::Abi::Scalar(callee)) => {
+                    caller.value == callee.value
+                }
+                (
+                    abi::Abi::ScalarPair(caller1, caller2),
+                    abi::Abi::ScalarPair(callee1, callee2),
+                ) => caller1.value == callee1.value && caller2.value == callee2.value,
+                // Be conservative
+                _ => false,
+            }
+        };
+        // Padding must be fully equal.
+        let pad_compat = || {
+            if caller_abi.pad != callee_abi.pad {
+                trace!(
+                    "check_argument_compat: incompatible pad: {:?} != {:?}",
+                    caller_abi.pad,
+                    callee_abi.pad
+                );
+                return false;
+            }
             return true;
-        }
-        if !rust_abi {
-            // Don't risk anything
-            return false;
-        }
-        // Compare layout
-        match (caller.abi, callee.abi) {
-            // Different valid ranges are okay (once we enforce validity,
-            // that will take care to make it UB to leave the range, just
-            // like for transmute).
-            (abi::Abi::Scalar(caller), abi::Abi::Scalar(callee)) => caller.value == callee.value,
-            (abi::Abi::ScalarPair(caller1, caller2), abi::Abi::ScalarPair(callee1, callee2)) => {
-                caller1.value == callee1.value && caller2.value == callee2.value
+        };
+        // For comparing the PassMode, we allow the attributes to differ
+        // (e.g., it is okay for NonNull to differ between caller and callee).
+        // FIXME: Are there attributes (`call::ArgAttributes`) that do need to be checked?
+        let mode_compat = || {
```

Comment:
I think there should be a strict mode option that is closer to the old behavior to catch cases where someone depends on details of the rust abi.

Reply:
The new behavior is strictly more strict than the old one, so... I am confused by this comment.

Reply:
The old behavior was to say that both are incompatible when the type doesn't match and either the rust abi isn't used or the rust abi used but it isn't an argument with the direct or pair pass mode. This is much stricter than considering the argument to be compatible when the pass modes match (minus arg attributes).

Reply:
For example the old code would consider `struct Foo(u8, u64, u8);` and `struct Bar(u64, u8, u8);` to be incompatible, but the new code will consider them compatible for the rust abi as both get a cast pass mode that passes a 64bit argument (for the `u64`) and a 16bit argument (packing both `u8` fields together).

Reply:
We still also compare the types -- I first thought comparing the modes would be enough but they do not contain remotely enough information (e.g., two `Indirect` arguments might have completely different types).

The idea is that the old type-based restrictions apply *plus* the mode has to be the same. Possibly the type-based restrictions could be relaxed (just requiring size and alignment to be the same, or even just the size) -- but so far I am not proposing to do that.

Reply:
> We still also compare the types

If the types are equal, it immediately says that it is equal. If the types are not equal, before this PR it would only consider it compatible in a limited amount of cases. After this PR the amount of cases is drastically increased. I would like a strict mode that only allows unequal types in a limited amount of cases just like before this PR.

before:
* types equal => definitively compatible
* types unequal =>
  * not rust abi => incompatible
  * rust abi =>
    * compatible scalar or scalar pair => compatible
    * else => incompatible

after:
* types equal => definitively compatible
* types unequal =>
  * compatible layout, padding and pass mode => compatible
  * else => incompatible

the strict mode I want would be something like:
* types equal => definitively compatible
* types unequal =>
  * not rust abi => compatible if compatible layout, padding and pass mode
  * rust abi => only compatible if layout, padding and pass mode are compatible and guaranteed to remain compatible forever and not just because we currently choose to do so.

Reply:
> After this PR the amount of cases is drastically increased.

For the Rust ABI, it is not. I am confused why you say that. The old conditions must be true *and* the pass mode must match.

Given that `FnAbi` does not contain the `Abi`, I also don't think we should make anything depend on the `Abi`.

Reply:
> It also leads to more code being accepted for rust_abi where the indirect and cast pass modes were previously always considered incompatible. In the rust abi it is unstable if an argument will be passed as indirect or cast. I can't think of any valid reasons to allow mismatched types for them even for libstd internal usage.

`layout_compat` implements exactly the old layout checks (except it ignores the `Abi`), and the requirement for compatibility is `layout_compat() && pad_compat() && mode_compat()`.

So, no, what you say is just not true I think. Previously we did not even look at modes so indirect and cast could easily have been considered compatible!

Reply:
> In the rust abi it is unstable if an argument will be passed as indirect or cast. I can't think of any valid reasons to allow mismatched types for them even for libstd internal usage.

We accept transmutes between `repr(Rust)` types whose fields happen to be laid out in the right way. I think we likewise should accept function calls whose `FnAbi` happens to match. I think anything else would be (a) inconsistent, and (b) not even well-defined since we don't have a clear spec for what *is* guaranteed to hold true about the Rust ABI into the future.

Reply:
I wish I could replace `layout_compat` by something more principled that reflects whether under current rustc, the data would actually be passed properly, but I don't know nearly enough about all this FnAbi madness to implement such a check. Unfortunately, ignoring the layout entirely is wrong (`PassMode` does not capture enough information), and comparing it for full equality is also wrong. It doesn't look like we actually have an abstraction that properly represents if arguments are compatible. Based on what @eddyb had said in the past I *thought* `FnAbi` provides that but now I don't think this is the case.

Reply:
Specifically, I am very confused about two points:
- Your summary even says that after this PR, for the Rust ABI, we accept strictly less than before. Before it was 'types equal or compatible scalar or scalar pair', now it is 'layout, padding, and pass mode compatible'. Here, 'layout compatible' is defined as 'types equal or compatible scalar or scalar pair', i.e., exactly the old definition. Do we agree so far? (I am assuming that equal types implies equal pass mode and padding; if that is not the case your summary is incorrect.)<br> So I don't see why it would make any sense to introduce an *even stricter mode* with this PR. (I also don't think we should even have that mode, or at least I don't want to be the one implementing it -- but even leaving that aside, it is orthogonal to this PR.)
- It *is* true that this PR accepts some calls that were previously rejected, namely calls for non-Rust ABIs where the types are different but the layout is compatible *and* the padding and pass mode match. This might be worth debating (it is the only change here that I see as potentially contentious), but I don't see a good reason to single out the Rust ABI now that we have pass mode information available. From what I understand, however, your 'strict mode' is *not* about this case.